src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -425,32 +426,50 @@ private enum AssignmentQueryKind {
 
 					return AssignmentQueryKind.Hopeless;
 
-				default:
-					var model = m_compilation.GetSemanticModel( assignment.Expression.SyntaxTree );
-					var typeInfo = model.GetTypeInfo( assignment.Expression );
-
-					// Type can be null in the case of an implicit conversion where the
-					// expression alone doesn't have a type. For example:
-					//   int[] foo = { 1, 2, 3 };
-					var typeToCheck = typeInfo.Type ?? typeInfo.ConvertedType;
-
-					if( assignment.Expression is BaseObjectCreationExpressionSyntax _ ) {
-						// When we have a new T() we don't need to worry about the value
-						// being anything other than an instance of T.
-						query = new ImmutabilityQuery(
-							ImmutableTypeKind.Instance,
-							typeToCheck
-						);
-					} else {
-						// In general we need to handle subtypes.
-						query = new ImmutabilityQuery(
-							ImmutableTypeKind.Total,
-							typeToCheck
-						);
+				case SyntaxKind.InvocationExpression:
+
+					// Some methods are known to have return values that are
+					// immutable (such as Enumerable.Empty()).
+					// These should be considered immutable by the Analyzer.
+					if( m_compilation
+						.GetSemanticModel( assignment.Expression.SyntaxTree )
+						.GetSymbolInfo( assignment.Expression )
+						.Symbol is not IMethodSymbol methodSymbol ) {
+						break;
+					}
+
+					if( m_context.IsReturnValueKnownToBeImmutable( methodSymbol ) ) {
+						return AssignmentQueryKind.NothingToCheck;

```

Comment:
Thinking out loud, we could probably approach this a little differently, such that this becomes `AssignmentQueryKind.MethodInvocationQuery` or some such, so that this `GetQueryForAssignment` is doing fewer checks, but I think this is fine & simple, and we can refactor as desired any time later.